### ISSUE TITLE
additional URL meta data

### ISSUE THREAD

**Issue by kurtseifried** [distributedweaknessfiling/dwf-workflow#34]
As per twitter:

https://twitter.com/joshbressers/status/1386775378878943242
This will be pretty easy to do. Today the references don't differentiate this.
I'm thinking maybe a refsource of "FIXED" and "INTRODUCED"

https://twitter.com/kees_cook/status/1386762231862362115
Nice! What's the best way to split "commits that fix this" from "references" and to add "commits that introduced the flaw"?
The Ubuntu CVE tracker regularly includes the "commits that introduced the flaw", which is very handy for validating -stable trees, doing forensics, etc.

So at a minimum it looks like we need "FIXED" and "INTRODUCED"

Do we also want to give hints about what a URL is, e.g. mailing list archive? web page? code patch? hashes of the content (e.g. code patches would benefit from this).

**Comment by kees** [comment]
I shy away from using URLs for SCM IDs -- they require additional parsing to map into the SCM. E.g. multiple URLs might point to the same git SHA (e.g. on GitHub, on kernel.org).

Perhaps add a "project SCM" URI (e.g. "git://git.kernel.org/..."), and then "commits", which could have "introduction" and "fix" (which both may be repeated: sometimes multiple commits are needed to fix a flaw).

**Comment by kurtseifried** [comment]
So one challenge is myself (as a former CVE board member) was stuck in the past, e.g. the fact that the database could only support a text description and URL's prior to my JSON format, which in turn only took the existing data (and even on this I got pushback). 

So let's open it up. URLs+meta data, cool

SCM data, what should that look like? Hash value obviously, ideally URL's to find it easily, what it applies to? author? commit message?

**Comment by kurtseifried** [comment]
Basically, why aren't we treating commits as first-class citizens? What else should be a first-class citizen?

**Comment by kees** [comment]
I think it depends on audience. For this kind of thing, I think there are three audiences:

- end users ("is my installation vulnerable to this?")
- developers ("is this fork of the source vulnerable to this?")
- researchers ("what sort of flaw is this?")

End users want to check versions (that can be done by a machine, "versions from v5.8 to before v5.11.16" is not very friendly, and is unclear about if the v5.9, v5.10 branches are vulnerable)

Developers want to see where a flaw was introduced and where it was fixed. (so, commit IDs)

Researchers want the same as developers (to study the code), and other classifiers to help narrow bulk analysis (e.g. CWEs), besides prose from a description, which tends to be pretty thin on details.

**Comment by kurtseifried** [comment]
Agreed, also there is the mix of end-user/developer, e.g. they have a custom web app that pulls stuff in, they can update it, but they also are somewhat dependant on upstream fixing it first. Another dimension of this data to me is:

Can I fix it? E.g. Chrome/Firefox are OpenSource but I can't easily build a new version and deploy it. I need to wait for upstream to fix it and ship it. Ditto for NodeJS, sure I can build a custom package, change all the deps, etc, but that level of work/expertise just isn't possible for most people. 

So for the "no I can't fix it" knowing that you're vulnerable (e.g. embedded library), and knowing when a fix is shipped and available is critical. In theory, this is what the AFFECTED data was meant to serve, but it clearly hasn't worked out, perhaps by allowing people to more easily update it the data will become available for use.

Do we perhaps need to express versions like 5.11.2 through 5.11.9 are affected, 5.11.10 has commit [hash value] that fixes it, which means a "version" is now potentially more than one piece of discrete information, do we just overload the text "5.11.10 has commit [hash value]" or do we split it up into a more structured way e.g.:

"5.11.10" AND "commit [hash value]" IS "fixed"

also it occurs to me in writing this that there is:

vulnerable
not vulnerable
fixed
introduced

e.g. not vulnerable and fixed are two very different states. I suspect vulnerable and introduced are different enough we might want them separated. Thoughts/comments?
